fq_test.go

```diff
@@ -0,0 +1,176 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"runtime"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// adapted from https://github.com/tadglines/wfq/blob/master/wfq_test.go
+
+type flowDesc struct {
+	// In
+	ftotal float64 // Total units in flow
+	imin   float64 // Min Packet servicetime
+	imax   float64 // Max Packet servicetime
+
+	// Out
+	idealPercent  float64
+	actualPercent float64
+}
+
+func genFlow(fq *FQScheduler, desc *flowDesc, key int) {
+	for i, t := 1, 0; t < int(desc.ftotal); i++ {
+		it := new(Packet)
+		it.queueidx = key
+		if desc.imin == desc.imax {
+			it.servicetime = desc.imax
+		} else {
+			it.servicetime = desc.imin + rand.Float64()*(desc.imax-desc.imin)
+		}
+		if float64(t)+it.servicetime > desc.ftotal {
+			it.servicetime = desc.ftotal - float64(t)
+		}
+		t += int(it.servicetime)
```

Comment:
This is unnecessarily casting away the fractional part.  Better to make `t` a `float64`.  For example, you could make the loop start with `for i, t := 1, float64(0)`.

Reply:
updated